rpmlint/checks/FileDigestCheck.py

```diff
@@ -34,46 +38,90 @@ def _get_digest_group_from_path(self, path):
-            elif known_digests and filename in known_digests:
-                expected_digests = known_digests[filename]
-                if 'skip' in expected_digests:
-                    pass
-                else:
-                    h = hashlib.new('sha256')
-                    with open(pkgfile.path, 'rb') as fd:
-                        while True:
-                            chunk = fd.read(4096)
-                            if not chunk:
-                                break
-                            h.update(chunk)
-                    hexdigest = h.hexdigest()
-                    if hexdigest not in expected_digests:
-                        self.output.add_info('E', pkg, f'{group}-file-digest-mismatch', filename,
-                                             f'expected:{",".join(expected_digests)}', f'has:{hexdigest}')
-            elif group:
-                self.output.add_info('E', pkg, f'{group}-file-digest-unauthorized', filename)
+        digest_hash = digest['hash']
+        pair = (pkgfile.name, algorithm)
+        if pair not in self.digest_cache:
+            h = hashlib.new(algorithm)
+            with open(pkgfile.path, 'rb') as fd:
+                while True:
+                    chunk = fd.read(4096)
+                    if not chunk:
+                        break
+                    h.update(chunk)
+            self.digest_cache[pair] = h.hexdigest()
+
+        return self.digest_cache[pair] == digest_hash
+
+    def _check_group_digests(self, digest_group, pkg):
+        for filename, digest in digest_group.items():
+            if not self._is_valid_digest(pkg.files[filename], digest):
+                return False
+        return True
+
+    def check_binary(self, pkg):
+        """
+        Check that all files in secured locations are covered by a file digest group
+        in which all files have valid digest.
+        """
+
+        if not self._check_symlinks(pkg):
+            return
+
+        # First collect all files that are in a digest configuration group
+        secured_paths = {pkgfile.name for pkgfile in pkg.files.values() if self._in_digest_configuration_group(pkgfile)}
+
+        # Iterate all digest groups and find one that covers all secured files
+        # and in which all digests match
+        for digest_group in self.digest_groups:
+            covered_files = set(digest_group.keys())
+            if secured_paths - covered_files:
+                # bail out as some secured paths are not covered
+                continue
+            elif covered_files - set(pkg.files.keys()):
+                # a group has a digest for a file that is not in the package
+                continue
+            elif self._check_group_digests(digest_group, pkg):
+                # we are done, we found a valid group
+                return
+
+        # Report errors
+        for filename in sorted(secured_paths):
+            group = self._get_digest_location_from_path(filename)
```

Comment:
This is about the fourth time that the group for a path is calculated. First in `_check_symlinks()` (for all files of the package) then again in `_in_digest_configuration_group()` (for all files of the package), then again in `secured_paths = ...` (for all files of the package) and then finally here, only for the `secured_paths`. I think this could be streamlined some more.

Reply:
Yep, I've added cache for it and renamed the function.